features/nanostack/mbed-mesh-api/source/wisun_tasklet.c

```diff
@@ -0,0 +1,319 @@
+ * @param event, describes the sender, receiver and event type.
+ *
+ * NOTE: Interrupts requested by HW are possible during this function!
+ */
+static void wisun_tasklet_main(arm_event_s *event)
+{
+    arm_library_event_type_e event_type;
+    event_type = (arm_library_event_type_e) event->event_type;
+
+    switch (event_type) {
+        case ARM_LIB_NWK_INTERFACE_EVENT:
+            /* This event is delivered every and each time when there is new
+             * information of network connectivity.
+             */
+            wisun_tasklet_parse_network_event(event);
+            break;
+
+        case ARM_LIB_TASKLET_INIT_EVENT:
+            /* Event with type EV_INIT is an initializer event of NanoStack OS.
+             * The event is delivered when the NanoStack OS is running fine.
+             * This event should be delivered ONLY ONCE.
+             */
+            mesh_system_send_connect_event(wisun_tasklet_data_ptr->tasklet);
+            break;
+
+        case ARM_LIB_SYSTEM_TIMER_EVENT:
+            eventOS_event_timer_cancel(event->event_id, wisun_tasklet_data_ptr->tasklet);
+
+            if (event->event_id == TIMER_EVENT_START_BOOTSTRAP) {
+                tr_debug("Restart bootstrap");
+                wisun_tasklet_configure_and_connect_to_network();
+            }
+            break;
+
+        case APPLICATION_EVENT:
+            if (event->event_id == APPL_EVENT_CONNECT) {
+                wisun_tasklet_configure_and_connect_to_network();
+            }
+            break;
+
+        default:
+            break;
+    } // switch(event_type)
+}
+
+/**
+ * \brief Network state event handler.
+ * \param event show network start response or current network state.
+ *
+ * - ARM_NWK_BOOTSTRAP_READY: Save NVK persistent data to NVM and Net role
+ * - ARM_NWK_NWK_SCAN_FAIL: Link Layer Active Scan Fail, Stack is Already at Idle state
+ * - ARM_NWK_IP_ADDRESS_ALLOCATION_FAIL: No WS Router at current Channel Stack is Already at Idle state
+ * - ARM_NWK_NWK_CONNECTION_DOWN: Connection to Access point is lost wait for Scan Result
+ * - ARM_NWK_NWK_PARENT_POLL_FAIL: Host should run net start without any PAN-id filter and all channels
+ * - ARM_NWK_AUHTENTICATION_FAIL: Pana Authentication fail, Stack is Already at Idle state
+ */
+static void wisun_tasklet_parse_network_event(arm_event_s *event)
+{
+    arm_nwk_interface_status_type_e status = (arm_nwk_interface_status_type_e) event->event_data;
+    tr_debug("app_parse_network_event() %d", status);
+    switch (status) {
+        case ARM_NWK_BOOTSTRAP_READY:
+            /* Network is ready and node is connected to Access Point */
+            if (wisun_tasklet_data_ptr->tasklet_state != TASKLET_STATE_BOOTSTRAP_READY) {
+                tr_info("Wi-SUN bootstrap ready");
+                wisun_tasklet_data_ptr->tasklet_state = TASKLET_STATE_BOOTSTRAP_READY;
+                wisun_tasklet_network_state_changed(MESH_CONNECTED);
```

Comment:
The local ula prefix needs probably to be checked similar to thread_tasklet.c to prevent reporting NSAPI_GLOBAL_UP with only ula prefix.

Reply:
@juhhei01 @mikter please comment

Reply:
Any comments about this? 6lowpan isn't doing it....

Reply:
@kjbracey-arm can you comment on this?

Reply:
I don't believe Wi-SUN has any inherent concept analogous to the "mesh-local" address.  Thread only says "local" if it has specifically only *the Thread mesh-local address*, not any ULA.

I believe the logic was that a standard ULA could equally represent a connection tunnel to a cloud server as well as a site-specific ULA, so we treat ULAs as "global" - we've no idea how many networks we can get with them. There could even be NAT going on, so we have a ULA but NATted global connectivity.

Apps can check to see if "get_ip_address" returns a ULA themselves if they want to distinguish between a ULA or need a real global address.

However, all the IPv6 things should be saying "local" if they only have a link-local address. The Thread stuff is just stretching that to also cover the thread mesh-local link.

Maybe if there was some other was to detect definitely local connectivity? Both 6LoWPAN-ND and Wi-SUN could maybe look at the "grounded" flag of the DODAG. If that was exposed via API, you would be local if clear. But then you'd need an event mechanism for it changing. Would be an extension.

Reply:
In IPv4, the global is actualy "We have a router and DHCP somewhere" so I think equal state should be indicated here.
I don't know how...

Do we have any API to ask that "is border router alive and does it have a backbone"

Because, the whole distinction between GLOBAL and LOCAL should not be about IP address scope. It should be answer to the question whether there is a route to "somewhere"
yes -> global
no -> local

And that information can be used in Mbed Client: "Should I try the TLS handshake again?" and the answer is the same.
GLOBAL -> yes
other states -> no

So there will most probably be cases where ULA prefix satisfies the requirement.

Reply:
Not sure we have any existing logic about route existence, or at least don't recall seeing it. 

I think maybe it's just convenient that having an IPv4 address implies a DHCPv4 server, and one assumes that a DHCPv4 server is a router.  We assume we don't auto-assign ourselves a "link-local" 169.254.x.x address.

So, IPv6 for LwIP checks for non-link-local IPv6 addresses. As does MeshInterfaceNanostack. "MESH_CONNECTED" is turned into either LOCAL_UP or GLOBAL_UP depending on if we only have a link-local address.

The Thread special case is stopping the mesh-local ULA from being reported as global, as any other ULA is.

So I think this is correct for now - it's plausible to assume if you have a ULA you at least have a router giving out SLAAC prefixes, even if it's not connected to anywhere, just like you assume the same for DHCPv4.

> Do we have any API to ask that "is border router alive and does it have a backbone"

Not sure whether it's easily or at all findable via public APIs, but RPL Grounded flag would be it, if the border router is maintaining it correctly. I think that's a general extension for 6LoWPAN-ND and Wi-SUN though - not something for this PR.